### DISCUSSION TITLE
Get the KClass of a KoClassDeclaration

### DISCUSSION THREAD

**Discussion by MrDienns** [LemonAppDev/konsist#731] [Ask a Question] [answered]
### Wanted end result.

For context: I am trying to utilize Konsist to assert all classes with a given parent are registered in a Kotlin serializer module (for polymorphic serialization). To achieve this, my first idea was to get all KClasses and check if they are part of a list somewhere. However, I am unable to figure out how to go from a KoClassDeclaration to a KClass (if it's even possible at all).

I am aware that this is a potential XY problem, so I'm open to suggestions on how to tackle this differently.

Here's what I got so far;
```kotlin
class ModuleTest {

    @Test
    fun `Should have all ConstructorCommandMessage instances registered in the serializer module`() {
        Konsist.scopeFromProject()
            .classes()
            .withAllParentsOf(ConstructorCommandMessage::class)
            .assertTrue {
                it.hasAnnotationOf(Serializable::class)
            }
    }
}
```

### What you tried so far.

I've looked through the Konsist source but was unable to find how to get the underlying KClass. I am already asserting that each discovered class has the `@Serializable` annotation, but with polymorphic serialization, this is assertion alone is not enough.

**Comment by MrDienns** [comment, marked as answer]
Managed to figure something out with the help of a friend, sharing here in case anyone else randomly stumbles upon this with the same or a similar question. We ended up just loading the KClass via `Class.forName` by passing in the fully qualified name that Konsist provides.

```kotlin
class ModuleTest {

    @Test
    fun `Should have all ConstructorCommandMessage instances registered in the serializer module`() {
        Konsist.scopeFromProject()
            .classes()
            .withParentNamed("CommandMessage.Complex", "CommandMessage.Atomic")
            .withAnnotationOf(Serializable::class)
            .assertTrue { shouldBeRegisteredInSerializer(CommandMessage::class, it.fullyQualifiedName) }
    }

    @OptIn(ExperimentalSerializationApi::class, InternalSerializationApi::class)
    private fun shouldBeRegisteredInSerializer(clazz: KClass<*>, fullyQualifiedClassName: String): Boolean {
        val configuration = buildSerializerModule()

        val kotlinClass = Class.forName(fullyQualifiedClassName).kotlin
        val serializer = kotlinClass.serializer()

        val registered = configuration.getPolymorphic(clazz, serializer.descriptor.serialName)
        return registered != null
    }

    private fun buildSerializerModule(): SerializersModule {
        return SerializersModule {
            include(DomainModule.build().serializersModule)
            include(DomainTestModule.build().serializersModule)
        }
    }
}
```

**Comment by igorwojda** [comment]
We are working on this. It will be released in the next release.

**Comment by nataliapeterwas** [comment]
A new version v0.14.0 has been released. It contains the fix.
See [v0.14.0 release](https://github.com/LemonAppDev/konsist/releases/tag/v0.14.0)

```kotlin
Konsist.scopeFromProject()
    .classes()
    .withAllParentsOf( ConstructorCommandMessage::class, indirectParents = true)
	.assertTrue {
		it.hasAnnotationOf(Serializable::class)
	}
```
